File: 2025/143. Reorder List/Solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public void reorderList(ListNode head) {
        ListNode dummy = new ListNode(0), slow = head, fast = head;
        dummy.next = head;
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }
        Stack<ListNode> stack = new Stack<>();
        while (slow != null) {
            stack.push(slow);
            slow = slow.next;
        }
        ListNode cur = dummy.next;
        while (!stack.isEmpty() && cur != null) {
            ListNode n = stack.pop(), tmp = cur.next;
            n.next = cur.next;
            cur.next = n;
            cur = cur.next.next;
        }
        cur.next = null;
    }
}

File: 2025/143. Reorder List/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        list = []
        dummy = ListNode()
        cur = dummy
        while head:
            list.append(head)
            head = head.next
        for i in range(math.ceil(len(list)/2)):
            node = list[i]
            next_node = list[-(i+1)]
            tmp = node.next
            cur.next = node
            node.next = next_node
            next_node.next = None
            if cur.next:
                cur = cur.next.next
        return dummy.next